### PR TITLE
[chore] [cleanup]: pytest, pytorch new versions, fix tests

### LINKED ISSUE #927: Compatibility with pytest 7
I just ran into an error in the test suite with the very recently release pytest 7 in https://github.com/conda-forge/fairscale-feedstock/pull/9.

I was too slow to copy out the log before azure deleted the run, but it was something about using assert in a boolean context or similar. After pinning to `pytest <7` the tests pass again for fairscale 0.4.5. Would be good if this could be fixed. :)

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?
* updated pytest to 7.0
* updated pytorch versions
* fixed a few tests
* added disable-tf32 code to some tests

Fixes #927 #921 .

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -161,7 +161,7 @@ run_oss_benchmark: &run_oss_benchmark
       name: Run OSS Benchmark
       command: |
         python benchmarks/oss.py --world_size 4 --epochs 2
-        python benchmarks/oss.py --check_regression --world_size 4 --optim_type oss_sharded_ddp
+        python benchmarks/oss.py --check_regression --world_size 4 --optim_type oss_sharded_ddp --epochs 12
```
Comment (min-xu-ai) [inline .circleci/config.yml:164]: added this so that we run 12 instead of 10 epochs to reduce the variability. runtime is not significantly longer.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -2309,6 +2312,20 @@ def _fsdp_instances(self) -> List["FullyShardedDataParallel"]:
         return [m for m in self.modules() if isinstance(m, FullyShardedDataParallel)]
 
     def _remove_uncollectable_params_from_optim_state_dict(self, osd: Dict) -> Dict:
+        """Return a new state dict filtering out the ones like MoE layers, which has
+        ``no_broadcast_optim_state`` flag set.
+
+        We also make rooms for the optimizer state on rank 0.
+        """
+        # In PyTorch version 1.12, Adam's `step` state changed from an into to a singleton
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:2320]: this is one of the fix for optimizer state handling. @sshleifer

File: fairscale/utils/testing.py [hunk 3]
```diff
@@ -136,6 +136,15 @@ def get_smi_ver() -> str:
     return tuple(int(n) for n in numbering)
 
 
+def make_cudnn_deterministic() -> None:
```
Comment (min-xu-ai) [inline fairscale/utils/testing.py:139]: added this function for tf32 disabling.

File: requirements-dev.txt [hunk 4]
```diff
@@ -14,9 +14,9 @@ mypy == 0.910
 pre-commit >= 2.15.0
 
 # Tools for unit tests & coverage.
-pytest == 5.4.1
-pytest-cov == 2.10.0
-pytest-timeout == 1.4.2
+pytest == 7.0.0
```
Comment (min-xu-ai) [inline requirements-dev.txt:17]: these are pytest upgrade

File: tests/experimental/nn/test_multiprocess_pipe.py [hunk 5]
```diff
@@ -103,20 +104,23 @@ def create(devices):
 
 
 @rpc_test()
+@skip_if_single_gpu
```
Comment (min-xu-ai) [inline tests/experimental/nn/test_multiprocess_pipe.py:107]: these are minor test fixes.

File: tests/nn/data_parallel/test_fsdp.py [hunk 6]
```diff
@@ -801,8 +801,11 @@ def __init__(self, group, wrapper_config, checkpoint_act=False, delay_before_fre
 
         if wrapper_config is not None:
             # we create a process group of size 1 for the expert params
-            expert_group = torch.distributed.new_group([group.rank()])  # world size 1 means no shard
-            expert = FullyShardedDataParallel(expert, expert_group, **wrapper_config)
+            # we also need to pass that group as the reduce_scatter group.
+            expert_group = torch.distributed.new_group([group.rank()])
+            expert = FullyShardedDataParallel(
+                expert, process_group=expert_group, process_group_reduce_scatter=expert_group, **wrapper_config
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp.py:807]: this change to avoid a warning from fdsp's init function due to reduce_scatter group.

File: tests/nn/data_parallel/test_fsdp_grad_acc.py [hunk 7]
```diff
@@ -64,6 +64,7 @@ def _test_nested_wrapper(self, rank, group, config):
 
     @classmethod
     def _test_grad_acc(self, model, batch_dim, use_no_sync_context=True):
+        make_cudnn_deterministic()
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_grad_acc.py:67]: made this test disable tf32. There are likely more of such tests. But I only have a single 3090 machine for testing this. will try to get an a100 later for testing.

File: tests/nn/data_parallel/test_fsdp_optimizer_utils.py [hunk 8]
```diff
@@ -24,11 +24,13 @@
 )
 
 
-def first_tensor_numel(dct):
+def all_tensors_numel_except_for_step(dct):
```
Comment (min-xu-ai) [inline tests/nn/data_parallel/test_fsdp_optimizer_utils.py:27]: this fixes this test so that new behavior of Adam is tested.

File: tests/nn/model_parallel/test_layers.py [hunk 9]
```diff
@@ -136,7 +134,7 @@ def run_test_initialize_affine_weight(rank, model_parallel_size, filename, filen
             torch.distributed.get_rank(), error
         )
     )
-    assert error < 1.0e-6
+    assert error < 1.0e-6, error
```
Comment (min-xu-ai) [inline tests/nn/model_parallel/test_layers.py:137]: these are for debugging help.

File: tests/nn/moe/test_moe_layer.py [hunk 10]
```diff
@@ -68,6 +69,7 @@ def expert_params(device):
 @pg_test()
 @pytest.mark.parametrize("device", devices)
 def forward(device):
+    make_cudnn_deterministic()
```
Comment (min-xu-ai) [inline tests/nn/moe/test_moe_layer.py:72]: make this test work on a100.

File: tests/nn/moe/test_top2gating.py [hunk 11]
```diff
@@ -27,17 +27,17 @@ def do_test_forward(device):
     gate = Top2Gate(4, 6).to(device)
     capacity = 2 * 12 // 6
     l_aux, combine_weights, dispatch_mask = gate(input)
-    assert pytest.approx(l_aux.item(), 0.0283)
+    assert pytest.approx(l_aux.item(), rel=0.01) == 0.0267, l_aux
```
Comment (min-xu-ai) [inline tests/nn/moe/test_top2gating.py:30]: the old code seems to use pytest.approx incorrectly and new version detects this now. The numerical value that was used was in fact incorrect. Hence the change of the value.

File: tests/optim/test_layerwise_gradient_scaler.py [hunk 12]
```diff
@@ -70,6 +70,7 @@ def load_data(model_type: str) -> Union[DataLoader, Tuple[Any, Any]]:
         torch.manual_seed(10)
         transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
 
+        # TODO: we should NOT do this download over and over again during test.
```
Comment (min-xu-ai) [inline tests/optim/test_layerwise_gradient_scaler.py:73]: this test is already disabled, flagging the download to be fixed.

File: benchmarks/oss.py [hunk 13]
```diff
@@ -264,14 +266,21 @@ def run_closure(closure, scaler, optimizer):
...
+                f"Epoch {epoch} - processed {measurements[-1]:.2f} img per sec. "
+                f"Loss {final_loss:.3f} min loss {min_loss:.3f}"
+            )
 
     training_stop = time.monotonic()
     img_per_sec = n_items / (training_stop - training_start) * args.epochs
     logging.info(f"[{dist.get_rank()}] : Training done. {img_per_sec:.2f} img per sec inc. checkpoint")
 
-    validate_benchmark(measurements, final_loss, args, check_regression)
+    # Use min_loss to check instead of final_loss since the final_loss is a bit random.
```
Comment (blefaudeux) [inline benchmarks/oss.py:280]: the final loss is a bit random ? it used to be reproducible, that's new

File: benchmarks/oss.py [hunk 14]
```diff
@@ -109,7 +109,8 @@ def validate_benchmark(measurements, final_loss, args, check_regression):
         assert max_memory < 1.05 * golden_data["reference_memory"], (
             f"Memory use regression detected: " f"{max_memory} vs. {1.05* golden_data['reference_memory']}"
         )
-        assert abs(cast(float, final_loss) - golden_data["reference_loss"]) < 1e-2, (
+        # any min_loss < than golden + epsilon is OK.
```
Comment (blefaudeux) [inline benchmarks/oss.py:112]: I know it's being discussed, but basically OSS is in pytorch already and FSDP is a superset of shardedddp feature-wise, so this could also be removed altogether to cut on the maintenance costs

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (blefaudeux) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:2320]: [nit] `step` state changed from an *into*\int to

File: benchmarks/oss.py (hunk 14 above)
Comment (min-xu-ai) [reply benchmarks/oss.py:112]: Agreed. I think we will have a cleanup change separately. I am in favor of retiring them soon.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:2320]: good catch! Thanks!

Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:2320]:
```suggestion
        # In PyTorch version 1.12, Adam's `step` state changed from an int to a singleton
```

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 15]
```diff
@@ -2309,6 +2312,20 @@ def _fsdp_instances(self) -> List["FullyShardedDataParallel"]:
         return [m for m in self.modules() if isinstance(m, FullyShardedDataParallel)]
 
     def _remove_uncollectable_params_from_optim_state_dict(self, osd: Dict) -> Dict:
+        """Return a new state dict filtering out the ones like MoE layers, which has
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:2315]: Where do we need this new function? Is this part of the tests refactoring change or something else?

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:2315]: This is not a new function. I added a comment

File: tests/optim/test_layerwise_gradient_scaler.py (hunk 12 above)
Comment (min-xu-ai) [reply tests/optim/test_layerwise_gradient_scaler.py:73]: Sure!

**Review by anj-s — APPROVED** [review]

**Review by tmarkstrum — APPROVED** [review]

**Comment by min-xu-ai** [comment]
@tmarkstrum, the reason seems to be related to randomness of the loss curve for this benchmarking training process. With more epochs, it is more likely that the min loss will be reaches, hence verify the module is likely correct.
